### PR TITLE
Add spec pipeline skeleton for wasp-config (app, page, query)

### PR DESCRIPTION
**Pull request by sodic**
## Description

The skeleton/tracer-bullet for the new App spec.

This PR covers `app`, `page`, `query`, and the analyzer that ties them together. I implemented it next to the existing legacy system and based its structure on what we already had.

The new spec API is not yet wired into the package. After we iterate on this PR, I plan to add more declarations. Once we have all of them, I'll flip the switch.

Advice for reviewing:
- Checkout the repo locally and compare with legacy files.
- Ignore the todos, they will be handled later.
- If you notice something that should have a todo, mention it.

## Type of change
that

- [x] **🔧 Just code/docs improvement** 
- [ ] **🐞 Bug fix** 
- [ ] **🚀 New/improved feature** 
- [ ] **💥 Breaking change** 

## Checklist

- [ ] I tested my change in a Wasp app to verify that it works as intended.

- 🧪 Tests and apps:

  - [x] I added **unit tests** for my change. 
  - [ ] _(if you fixed a bug)_ I added a **regression test** for the bug I fixed. 
  - [ ] _(if you added/updated a feature)_ I added/updated **e2e tests** in \`examples/kitchen-sink/e2e-tests\`.
  - [ ] _(if you added/updated a feature)_ I updated the **starter templates** in \`waspc/data/Cli/templates\`, as needed.
  - [ ] _(if you added/updated a feature)_ I updated the **example apps** in \`examples/\`, as needed.
    - [ ] _(if you updated \`examples/tutorials\`)_ I updated the tutorial in the docs (and vice versa).

- 📜 Documentation:

  - [ ] _(if you added/updated a feature)_ I **added/updated the documentation** in \`web/docs/\`.

- 🆕 Changelog: _(if change is more than just code/docs improvement)_
  - [ ] I updated \`waspc/ChangeLog.md\` with a **user-friendly** description of the change.
  - [ ] _(if you did a breaking change)_ I added a step to the current **migration guide** in \`web/docs/migration-guides/\`.
  - [ ] I **bumped the \`version\`** in \`waspc/waspc.cabal\` to reflect the changes I introduced.

### CODE REVIEW

File: waspc/data/packages/wasp-config/__tests__/spec/appAnalyzer.unit.test.ts
Comment (sodic) [inline waspc/data/packages/wasp-config/__tests__/spec/appAnalyzer.unit.test.ts:1]: @FranjoMindek These are all modeled after legacy tests. I suggest you check out the repo and compare them. You'll know what's up since you're the author.

File: waspc/data/packages/wasp-config/__tests__/spec/testFixtures.ts [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+      };
+    }
+    case "default":
+      return { importDefault: "defaultExport", from: "@src/external" };
+    default:
+      assertUnreachable(importKind);
+  }
+}
+
+function assertUnreachable(value: never): never {
```
Comment (sodic) [inline waspc/data/packages/wasp-config/__tests__/spec/testFixtures.ts:97]: All utilities below this point are the same as in legacy, no need to look at them.

File: waspc/data/packages/wasp-config/src/spec/appAnalyzer.ts
Comment (sodic) [inline waspc/data/packages/wasp-config/src/spec/appAnalyzer.ts:1]: Most of the code is similar to legacy, I recommend comparing them locally.

File: waspc/data/packages/wasp-config/src/spec/mapApp.ts [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+  return function parseRef(potentialRef: string): AppSpec.Ref<T> {
+    if (!declNames.includes(potentialRef)) {
+      throw new Error(`Invalid ${declType} reference: ${potentialRef}`);
+    }
+    return {
+      name: potentialRef,
+      declType,
+    };
+  };
+}
```
Comment (sodic) [inline waspc/data/packages/wasp-config/src/spec/mapApp.ts:126]: Same as in legacy.

File: waspc/data/packages/wasp-config/src/spec/mapTsAppSpecToAppSpecDecls.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+      name: potentialRef,
+      declType,
+    };
+  };
+}
+
+type GetPartForKind<Id extends TsAppSpec.Part["kind"]> = Extract<
+  TsAppSpec.Part,
+  { kind: Id }
+>;
```
Comment (sodic) [inline waspc/data/packages/wasp-config/src/spec/mapTsAppSpecToAppSpecDecls.ts]: Modeled after `GetDeclForType`.

File: waspc/data/packages/wasp-config/src/spec/mapApp.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+  id: Id,
+  parts: TsAppSpec.Part[],
+): GetPartForKind<Id>[] {
+  return parts.filter((p): p is GetPartForKind<Id> => p.kind === id);
+}
+
+function mapToDecls<T, DeclType extends AppSpec.Decl["declType"]>(
+  items: T[],
+  declType: DeclType,
+  deriveName: (item: T) => string,
```
Comment (sodic) [inline waspc/data/packages/wasp-config/src/spec/mapApp.ts:143]: This name derivation the only part that makes it different from legacy.

File: waspc/data/packages/wasp-config/src/spec/mapApp.ts [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+  mapValue: (item: T) => AppSpec.GetDeclForType<DeclType>["declValue"],
+) {
+  return items.map((item) => ({
+    declType,
+    declName: deriveName(item),
+    declValue: mapValue(item),
+  }));
+}
+
+export function deriveExtImportName(extImport: TsAppSpec.ExtImport): string {
```
Comment (sodic) [inline waspc/data/packages/wasp-config/src/spec/mapApp.ts:153]: This is a new thing and we need it because declarations no longer have names. Their names are derived.

File: waspc/data/packages/wasp-config/src/spec/mapApp.ts [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+    return extImport.alias ?? extImport.import;
+  }
+  return extImport.importDefault;
+}
+
+function makeDeclsArray(decls: {
+  [Type in AppSpec.Decl["declType"]]: AppSpec.GetDeclForType<Type>[];
+}): AppSpec.Decl[] {
+  return Object.values(decls).flatMap((decl) => [...decl]);
+}
```
Comment (sodic) [inline waspc/data/packages/wasp-config/src/spec/mapApp.ts:179]: Same as in legacy.

File: waspc/data/packages/wasp-config/src/spec/mapApp.ts [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+      kind: "default",
+      name: extImport.importDefault,
+      path: extImport.from,
+    };
+  } else {
+    throw new Error(
+      "Invalid ExtImport: neither `import` nor `importDefault` is defined",
+    );
+  }
+}
```
Comment (sodic) [inline waspc/data/packages/wasp-config/src/spec/mapApp.ts:107]: Same as in legacy.

File: waspc/data/packages/wasp-config/src/spec/publicApi/tsAppSpec.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
+export type TsAppSpec = {
+  kind: "app";
+  name: string;
+  wasp: { version: string };
+  title: string;
+  head?: string[];
+  parts: Part[];
+};
+
+export type Part = Page | Query;
```
Comment (FranjoMindek) [inline waspc/data/packages/wasp-config/src/spec/publicApi/tsAppSpec.ts:9]: 🥲

File: waspc/data/packages/wasp-config/src/spec/publicApi/tsAppSpec.ts [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+export type Query = {
+  kind: "query";
+  fn: ExtImport;
+  entities?: string[];
+  auth?: boolean;
+};
+
+export type ExtImport =
+  | { import: string; alias?: string; from: `@src/${string}` }
```
Comment (FranjoMindek) [inline waspc/data/packages/wasp-config/src/spec/publicApi/tsAppSpec.ts:29]: No `kind` here?

I guess that is:
- We only need `kind` on `Part`s, to differentiate them.
- It makes the API easier, users don't have to specify the export `kind`.

If so is true, `Part` should explain that any `Part` type requires a `kind`.
Or encode it into the type system.

File: waspc/data/packages/wasp-config/src/spec/mapApp.ts [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+  const entityRefParser = makeRefParser("Entity", entityNames);
+
+  const pages = extractParts("page", parts);
+  const pageDecls = mapToDecls(
+    pages,
+    "Page",
+    (page) => deriveExtImportName(page.component),
+    mapPage,
+  );
```
Comment (FranjoMindek) [inline waspc/data/packages/wasp-config/src/spec/mapApp.ts:26]: Use named parameters instead?

File: waspc/data/packages/wasp-config/src/spec/mapApp.ts [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+  const appDecl = {
+    declType: "App" as const,
+    declName: name,
+    declValue: {
+      wasp,
+      title,
+      head,
+      auth: undefined,
```
Comment (FranjoMindek) [inline waspc/data/packages/wasp-config/src/spec/mapApp.ts:44]:
```suggestion
      head,
      // TODO: add these guys
      auth: undefined,
```

File: waspc/data/packages/wasp-config/src/spec/mapTsAppSpecToAppSpecDecls.ts [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+      throw new Error(`Invalid ${declType} reference: ${potentialRef}`);
+    }
+    return {
+      name: potentialRef,
+      declType,
+    };
+  };
+}
+
+type GetPartForKind<Id extends TsAppSpec.Part["kind"]> = Extract<
```
Comment (FranjoMindek) [inline waspc/data/packages/wasp-config/src/spec/mapTsAppSpecToAppSpecDecls.ts]: Why `Id` over `Kind`?

File: waspc/data/packages/wasp-config/src/spec/mapTsAppSpecToAppSpecDecls.ts [hunk 13]
```diff
@@ -0,0 +1,160 @@
...
+  TsAppSpec.Part,
+  { kind: Id }
+>;
+
+function extractParts<Id extends TsAppSpec.Part["kind"]>(
+  id: Id,
```
Comment (FranjoMindek) [inline waspc/data/packages/wasp-config/src/spec/mapTsAppSpecToAppSpecDecls.ts]: Why `id`/`Id` over `kind`/`Kind`?

File: waspc/data/packages/wasp-config/src/spec/mapApp.ts (hunk 4 above)
Comment (FranjoMindek) [reply waspc/data/packages/wasp-config/src/spec/mapApp.ts:143]: Idea, if we are sure the name always come from the `ExtImport` (so far true):

Instead of accepting `deriveName` make it accept a `(item: T) => TsAppSpec.ExtImport` fn, and then inline the `deriveExtImportName` inside of `mapToDecls`.

Now the call looks like this:
```ts
const pages = extractParts("page", parts);
const pageDecls = mapToDecls(
  pages,
  "Page",
  (page) => page.component,
  mapPage,
);
```
But I wouldn't delve into reflection stuff (another step forward).

Personally I prefer the `deriveName`. If you too, sorry for wasting your time.

File: waspc/data/packages/wasp-config/src/spec/mapApp.ts [hunk 14]
```diff
@@ -0,0 +1,160 @@
...
+      server: undefined,
+      client: undefined,
+      db: undefined,
+      emailSender: undefined,
+      webSocket: undefined,
+    },
+  };
+
+  return makeDeclsArray({
```
Comment (FranjoMindek) [inline waspc/data/packages/wasp-config/src/spec/mapApp.ts:53]: Any reason not to: `return [appDecl, ...pageDecls, ...queryDecls]`?
All of the variables we have pass `satisfies AppSpec.Decl` or `satisfies AppSpec.Decl[]`.

File: waspc/data/packages/wasp-config/src/spec/appAnalyzer.ts [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+        "calling `app({ ... })`. Make sure you export that value.",
+    };
+  }
+
+  return { status: "ok", value: usersDefaultExport };
+}
+
+// TODO: This should probably live elsewhere
+function isTsAppSpec(value: unknown): value is TsAppSpec.TsAppSpec {
+  // TODO: Make this more robust, I'm not a fan of the hardcoded "kind" - neither here nor in other places
```
Comment (FranjoMindek) [inline waspc/data/packages/wasp-config/src/spec/appAnalyzer.ts]: Good question, can we use a library for this?
Some unknown object to type validation.

**Review by FranjoMindek — COMMENTED** [review]
Part 1/2, didn't do tests yet.
Might help you to work on this sooner.

File: waspc/data/packages/wasp-config/src/spec/mapApp.ts (hunk 4 above)
Comment (sodic) [reply waspc/data/packages/wasp-config/src/spec/mapApp.ts:143]: > Idea, if we are sure the name always come from the ExtImport (so far true):

Not necessarily, one example is route names which we'll probably base on paths.

But in any case, I'd keep generalisations modest for now. Patterns will emerge once we get more declarations in.

File: waspc/data/packages/wasp-config/src/spec/publicApi/tsAppSpec.ts (hunk 8 above)
Comment (sodic) [reply waspc/data/packages/wasp-config/src/spec/publicApi/tsAppSpec.ts:9]: Don't worry, the LLM does it (so :smiling_face_with_tear: indeed).

File: waspc/data/packages/wasp-config/src/spec/publicApi/tsAppSpec.ts (hunk 9 above)
Comment (sodic) [reply waspc/data/packages/wasp-config/src/spec/publicApi/tsAppSpec.ts:29]: That's a good point, I'll explain it in the code.

The only reason we actually need kinds is queries and actions. Because their structure is identical.
We might have other similar declarations in the future. So I thought better safe than sorry for now.

File: waspc/data/packages/wasp-config/src/spec/mapApp.ts (hunk 10 above)
Comment (sodic) [reply waspc/data/packages/wasp-config/src/spec/mapApp.ts:26]: image

Comment (sodic) [reply waspc/data/packages/wasp-config/src/spec/mapApp.ts:26]: Jokes aside, I'd leave the generalizations for later - this function will likely evolve. I'll add a note reminding me to go back to it.

File: waspc/data/packages/wasp-config/src/spec/mapTsAppSpecToAppSpecDecls.ts (hunk 12 above)
Comment (sodic) [reply waspc/data/packages/wasp-config/src/spec/mapTsAppSpecToAppSpecDecls.ts]: Leftover from when this was called `PartId`. I'll fix it.

File: waspc/data/packages/wasp-config/src/spec/mapApp.ts (hunk 14 above)
Comment (sodic) [reply waspc/data/packages/wasp-config/src/spec/mapApp.ts:53]: Yes, check the third point listed here: https://github.com/wasp-lang/wasp/pull/2393#discussion_r1866620833

I should explain this in a comment though, so nice catch.

File: waspc/data/packages/wasp-config/src/spec/appAnalyzer.ts (hunk 15 above)
Comment (sodic) [reply waspc/data/packages/wasp-config/src/spec/appAnalyzer.ts]: There possibly is, but we don't need a library, it's simple enough. And library would likely make it more complicated.

We just need to add some type declarations (e.g., `const discriminator: keyof TsAppSpec = "kind"` is probably good enough).

File: waspc/data/packages/wasp-config/src/spec/mapApp.ts (hunk 4 above)
Comment (FranjoMindek) [reply waspc/data/packages/wasp-config/src/spec/mapApp.ts:143]: Yep, same reason why I prefer the current `deriveName` for now.

File: waspc/data/packages/wasp-config/src/spec/appAnalyzer.ts (hunk 15 above)
Comment (FranjoMindek) [reply waspc/data/packages/wasp-config/src/spec/appAnalyzer.ts]: Depends on what we want to do.
My thought was to test the whole `TsAppSpec` schema is correct, `Part`s included.

If we only want to test the first layer of `TsAppSpec`, then yes, overkill.

File: waspc/data/packages/wasp-config/__tests__/spec/appAnalyzer.unit.test.ts [hunk 16]
```diff
@@ -0,0 +1,90 @@
...
+  });
+
+  test("should parse full app successfully", async () => {
+    await testAnalyzeApp({
+      spec: Fixtures.getApp("full"),
+      entities: Fixtures.getEntities("full"),
+    });
+  });
+
+  test("should parse spec from async default export", async () => {
```
Comment (FranjoMindek) [inline waspc/data/packages/wasp-config/__tests__/spec/appAnalyzer.unit.test.ts]: Can be done by `testAnalyzeApp` if you add `asyncDefaultExport` option.

File: waspc/data/packages/wasp-config/__tests__/spec/testFixtures.ts [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+        title: "Mock App",
+        parts: [],
+      });
+    case "full":
+      return app({
+        name: "FullApp",
+        wasp: { version: "^0.16.3" },
+        title: "Mock App",
+        head: ['<link rel="icon" href="/favicon.ico" />'],
+        // TODO: should we fill the parts here? The old approach didn't have this problem
```
Comment (FranjoMindek) [inline waspc/data/packages/wasp-config/__tests__/spec/testFixtures.ts]: I would say:
- The `full` App config should be a config with at least one member of each possible `Part`. That would go in line what `Full` meant so far.
- Maybe we don't need the "integration" tests at all. I was at odds was the legacy test an integration test too. We only ever tested the result of `mapTsAppSpecToAppSpecDecls`, so most likely not. I think Martin thought they are, and I wasn't so sure. I would mostly like collapse those two files into a single one.
- So ideally, App tests would test full/minimal + edge cases (error throws etc.).

**Review by FranjoMindek — COMMENTED** [review]
Test comments.

File: waspc/data/packages/wasp-config/src/spec/appAnalyzer.ts (hunk 15 above)
Comment (sodic) [reply waspc/data/packages/wasp-config/src/spec/appAnalyzer.ts]: Haskell validates most of this stuff through the app spec, and the typechecker enforces the structure in user files. I don't think we need more than a simple check, just enough to cover cases like "they forgot to export anything" or "they exported the wrong thing."

File: waspc/data/packages/wasp-config/__tests__/spec/appAnalyzer.unit.test.ts (hunk 16 above)
Comment (sodic) [reply waspc/data/packages/wasp-config/__tests__/spec/appAnalyzer.unit.test.ts]: I wouldn't do it. We're pollute a helper with an extra option for just a single test case which operates at a different level than all the others.

File: waspc/data/packages/wasp-config/__tests__/spec/testFixtures.ts (hunk 17 above)
Comment (sodic) [reply waspc/data/packages/wasp-config/__tests__/spec/testFixtures.ts]: I see what you mean. I'd say the old code did need an "integration" test because other declarations weren't nested under the app. There was an object collecting and orchestrating them behind the scenes, and we needed to test that orchestration.

With the new system, there is no orchestration and we don't need an integration test. So, great suggestion, I'll get it done :)

File: waspc/data/packages/wasp-config/src/spec/publicApi/constructors.ts [hunk 18]
```diff
@@ -0,0 +1,19 @@
+import type { App, Page, Query } from "./tsAppSpec.js";
+
+export function app(input: Omit<App, "kind">): App {
+  return input;
+}
+
+export function page(
+  component: Page["component"],
+  config?: Pick<Page, "authRequired">,
```
Comment (infomiho) [inline waspc/data/packages/wasp-config/src/spec/publicApi/constructors.ts:9]: Isn't it better to go with `Omit` here? I think this way there will be less need to edit this type if we add new properties to `config` part of the declaration.

**Review by infomiho — APPROVED** [review]
Code-wise, clear, but there are decisions to be made about naming

File: waspc/data/packages/wasp-config/src/spec/publicApi/constructors.ts (hunk 18 above)
Comment (sodic) [reply waspc/data/packages/wasp-config/src/spec/publicApi/constructors.ts:9]: I didn't use it on purpose. I wanted the public API to be explicit because that way:
- You see all the stuff that can go into `config` without "calculating" it in your head.
- Adding new stuff and moving it around does not change the public interface unless you change it explicitly.
